consensus/e2e_tests/utils_test.go

```diff
@@ -526,40 +536,328 @@ func baseRpcMock(t *testing.T, _ modules.EventsChannel) *mockModules.MockRPCModu
+	require.NoError(t, err)
+	for pocketId, pocketNode := range pocketNodes {
+		nodeState := GetConsensusNodeState(pocketNode)
+		// Leader has already committed the block and hence moved to the next height.
+		if pocketId == leaderId {
+			assertNodeConsensusView(t, pocketId,
+				typesCons.ConsensusNodeState{
+					Height: height + 1,
+					Step:   1,
+					Round:  round,
+				},
+				nodeState)
+			require.Equal(t, typesCons.NodeId(0), nodeState.LeaderId, "Leader should be empty")
+			continue
 		}
-	}).AnyTimes()
+		assertNodeConsensusView(t, pocketId,
+			typesCons.ConsensusNodeState{
+				Height: height,
+				Step:   step,
+				Round:  round,
+			},
+			nodeState)
+		require.Equal(t, leaderId, nodeState.LeaderId, fmt.Sprintf("%d should be the current leader", leaderId))
+	}
+
+	return decideProposal
+}
 
-	return stateMachineMock
+// waitForNodeToSync waits for a node to sync to a target height
+// For every missing block for the unsynced node:
+//
+//	first, waits for the node to request a missing block via `waitForNodeToRequestMissingBlock()` function,
+//	then, waits for the node to receive the missing block via `waitForNodeToReceiveMissingBlock()` function,
+//	finally, wait for the node to catch up to the target height via `waitForNodeToCatchUp()` function.
+func waitForNodeToSync(t *testing.T,
+	clck *clock.Mock,
+	eventsChannel modules.EventsChannel,
+	unsyncedNode *shared.Node,
+	allNodes IdToNodeMapping,
+	targetHeight uint64) {
+
+	consensusMod := unsyncedNode.GetBus().GetConsensusModule()
+	currentHeight := consensusMod.CurrentHeight()
+
+	for i := currentHeight; i <= targetHeight; i++ {
+		blockRequest, err := waitForNodeToRequestMissingBlock(t, clck, eventsChannel, allNodes, currentHeight, targetHeight)
+		require.NoError(t, err)
+
+		blockResponse, err := waitForNodeToReceiveMissingBlock(t, clck, eventsChannel, allNodes, blockRequest)
+		require.NoError(t, err)
+
+		err = waitForNodeToCatchUp(t, clck, eventsChannel, unsyncedNode, blockResponse, targetHeight)
+		require.NoError(t, err)
+	}
+}
+
+// TODO (#352): implement this function
+// waitForNodeToRequestMissingBlock waits for unsynced node to request missing block form the network
+func waitForNodeToRequestMissingBlock(
+	t *testing.T,
+	clck *clock.Mock,
+	eventsChannel modules.EventsChannel,
+	allNodes IdToNodeMapping,
+	startingHeight uint64,
+	targetHeight uint64) (*anypb.Any, error) {
+
+	return &anypb.Any{}, nil
+}
+
+// TODO (#352): implement this function.
+// waitForNodeToReceiveMissingBlock requests block request of the unsynced node
+// for given node to node to catch up to the target height by sending the requested block.
+func waitForNodeToReceiveMissingBlock(
+	t *testing.T,
+	clck *clock.Mock,
+	eventsChannel modules.EventsChannel,
+	allNodes IdToNodeMapping,
+	blockReq *anypb.Any) (*anypb.Any, error) {
```

Comment:
leaving one comment but ditto everywhere else

```suggestion
	blockReq *anypb.Any
) (*anypb.Any, error) {
```